libbeat/processors/timeseries/timeseries.go

```diff
@@ -0,0 +1,133 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package timeseries
+
+import (
+	"strings"
+
+	"github.com/mitchellh/hashstructure"
+
+	"github.com/elastic/beats/libbeat/asset"
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+type timeseriesProcessor struct {
+	dimensions map[string]interface{}
+	prefixes   map[string]interface{}
+}
+
+// NewTimeSeriesProcessor returns a processor to add timeseries info to events
+// Events are processed to extract all their dimensions (keyword fields that
+// hold a dimension of the metrics) and compute a hash of all their values into
+// `timeseries.instance` field.
+func NewTimeSeriesProcessor(beatName string) (processors.Processor, error) {
+	fieldsYAML, err := asset.GetFields(beatName)
+	if err != nil {
+		return nil, err
+	}
+
+	fields, err := common.NewFieldsFromYAML(fieldsYAML)
+	if err != nil {
+		return nil, err
+	}
+
+	dimensions := map[string]interface{}{}
+	prefixes := map[string]interface{}{}
+	populateDimensions("", dimensions, prefixes, fields)
+
+	return &timeseriesProcessor{dimensions: dimensions, prefixes: prefixes}, nil
+}
+
+func (t *timeseriesProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	if event.TimeSeries {
+		instanceFields := common.MapStr{}
+
+		// map all dimensions & values
+		for k, v := range event.Fields.Flatten() {
+			if t.isDimension(k) {
+				instanceFields[k] = v
+			}
+		}
+
+		h, err := hashstructure.Hash(instanceFields, nil)
```

Comment:
When you mentioned JSON it somehow reminded me of https://github.com/elastic/elasticsearch/issues/25312 but I don't think it would apply here as there is not such thing like "group by all these fields".

I'm less worried about polluting the document then about the size of it and if it would have some effect on the overall size. Probably not as the same would show up many times and potentially it even compresses better then a unique hash value. The reason I'm not concerned about polluting the document is that I do not expect that users often look at the raw document and mainly on the metrics. For debugging purpose I definitively think it would be beneficial to have the full string. In case things don't work as expected we would directly see what things are grouped by.